File: Providers/SqlDataProvider/SqlDataProvider.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports Microsoft.ApplicationBlocks.Data
Imports DotNetNuke

Namespace DNNStuff.SQLViewPro

    Public Class SqlDataProvider
        Inherits DataProvider

#Region "Private Members"
        Private Const ProviderType As String = "data"

        Private _providerConfiguration As Framework.Providers.ProviderConfiguration = Framework.Providers.ProviderConfiguration.GetProviderConfiguration(ProviderType)
        Private _connectionString As String
        Private _providerPath As String
        Private _objectQualifier As String
        Private _databaseOwner As String
#End Region

#Region "Constructors"

        Public Sub New()

            ' Read the configuration specific information for this provider
            Dim objProvider As Framework.Providers.Provider = CType(_providerConfiguration.Providers(_providerConfiguration.DefaultProvider), Framework.Providers.Provider)

            ' Read the attributes for this provider
            _connectionString = DotNetNuke.Common.Utilities.Config.GetConnectionString()

            _providerPath = objProvider.Attributes("providerPath")

            _objectQualifier = objProvider.Attributes("objectQualifier")
            If _objectQualifier <> "" And _objectQualifier.EndsWith("_") = False Then
                _objectQualifier += "_"
            End If

            _databaseOwner = objProvider.Attributes("databaseOwner")
            If _databaseOwner <> "" And _databaseOwner.EndsWith(".") = False Then
                _databaseOwner += "."
            End If

        End Sub

        Public ReadOnly Property ConnectionString() As String
            Get
                Return _connectionString
            End Get
        End Property

        Public ReadOnly Property ProviderPath() As String
            Get
                Return _providerPath
            End Get
        End Property

        Public ReadOnly Property ObjectQualifier() As String
            Get
                Return _objectQualifier
            End Get
        End Property

        Public ReadOnly Property DatabaseOwner() As String
            Get
                Return _databaseOwner
            End Get
        End Property
        Public ReadOnly Property CmdPrefix() As String
            Get
                Return "DNNStuff_SQLViewPro_"
            End Get
        End Property
#End Region

        Public Overrides Function RunQuery(ByVal queryText As String, ByVal dataSetName As String) As DataSet
            ' make replacements for objectQualifier and databaseOwner
            queryText = Text.RegularExpressions.Regex.Replace(queryText, "{oQ}|{objectQualifier}", ObjectQualifier, Text.RegularExpressions.RegexOptions.IgnoreCase)
            queryText = Text.RegularExpressions.Regex.Replace(queryText, "{dO}|{databaseOwner}", DatabaseOwner, Text.RegularExpressions.RegexOptions.IgnoreCase)
            Using cn As SqlConnection = New SqlConnection(ConnectionString)
                Using cmd As SqlCommand = SqlHelper.CreateCommand(cn, queryText, Nothing)
                    cmd.CommandTimeout = 0
                    cmd.CommandType = CommandType.Text
                    Dim adapter As SqlDataAdapter = New SqlDataAdapter(cmd)
                    cn.Open()
                    Dim ds As DataSet = New DataSet(dataSetName)
                    adapter.Fill(ds)
                    Return ds
                End Using
            End Using
        End Function

        ' report set
        Public Overrides Function GetReportSet(ByVal ReportSetId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetReportSet", ReportSetId), IDataReader)
        End Function
        Public Overrides Function GetReportSetByModule(ByVal ModuleId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetReportSetByModule", ModuleId), IDataReader)
        End Function
        Public Overrides Function UpdateReportSet(ByVal ModuleId As Integer, ByVal ReportSetId As Integer, ByVal ReportSetName As String, ByVal ReportSetTheme As String, ByVal ReportSetConnectionId As Integer, ByVal ReportSetHeaderText As String, ByVal ReportSetFooterText As String, ByVal ReportSetDebug As Boolean, ByVal RunCaption As String, ByVal BackCaption As String, ByVal ParameterLayout As String, ByVal AlwaysShowParameters As Boolean, ByVal RenderMode As String, ByVal AutoRun As Boolean, ByVal ReportSetConfig As String) As Integer
            Return CType(SqlHelper.ExecuteScalar(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "UpdateReportSet", ModuleId, ReportSetId, ReportSetName, ReportSetTheme, ReportSetConnectionId, ReportSetHeaderText, ReportSetFooterText, ReportSetDebug, RunCaption, BackCaption, ParameterLayout, AlwaysShowParameters, RenderMode, AutoRun, ReportSetConfig), Integer)
        End Function

        Public Overrides Function ListReportSet(ByVal PortalId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "ListReportSet", PortalId), IDataReader)
        End Function

        Public Overrides Sub DeleteReportSet(ByVal ModuleId As Integer)
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "DeleteReportSet", ModuleId)
        End Sub

        Public Overrides Function GetReportSetReport(ByVal ReportSetId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetReportSetReport", ReportSetId), IDataReader)
        End Function

        Public Overrides Function GetReportSetParameter(ByVal ReportSetId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetReportSetParameter", ReportSetId), IDataReader)
        End Function

        ' report
        Public Overrides Function GetReport(ByVal ReportId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetReport", ReportId), IDataReader)
        End Function

        Public Overrides Function UpdateReport(ByVal ReportSetId As Integer, ByVal ReportId As Integer, ByVal ReportTypeId As String, ByVal ReportName As String, ByVal ReportTheme As String, ByVal ReportConnectionId As Integer, ByVal ReportHeaderText As String, ByVal ReportFooterText As String, _
                ByVal ReportCommand As String, ByVal ReportConfig As String, ByVal ReportOrder As Integer, ByVal ReportDrilldownReportId As Integer, ByVal ReportDrilldownFieldname As String, ByVal ReportNoItemsText As String, ByVal ReportPageTitle As String, ByVal ReportCommandCacheTimeout As Integer, ByVal ReportMetaDescription As String) As Integer
            Return CType(SqlHelper.ExecuteScalar(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "UpdateReport", ReportSetId, ReportId, ReportTypeId, ReportName, ReportTheme, ReportConnectionId, ReportHeaderText, ReportFooterText, _
                ReportCommand, ReportConfig, ReportOrder, ReportDrilldownReportId, ReportDrilldownFieldname, ReportNoItemsText, ReportPageTitle, ReportCommandCacheTimeout, ReportMetaDescription), Integer)
        End Function

        Public Overrides Function ListReport(ByVal ReportSetId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "ListReport", ReportSetId), IDataReader)
        End Function

        Public Overrides Function ListReportDrilldown(ByVal ReportId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "ListReportDrilldown", ReportId), IDataReader)
        End Function

        Public Overrides Sub UpdateReportOrder(ByVal ReportId As Integer, ByVal Increment As Integer)
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "UpdateReportOrder", ReportId, Increment)
        End Sub
        Public Overrides Sub DeleteReport(ByVal ReportId As Integer)
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "DeleteReport", ReportId)
        End Sub

        ' Parameter
        Public Overrides Function GetParameter(ByVal ParameterId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetParameter", ParameterId), IDataReader)
        End Function

        Public Overrides Function UpdateParameter(ByVal ReportSetId As Integer, ByVal ParameterId As Integer, ByVal ParameterName As String, ByVal ParameterCaption As String, _
            ByVal ParameterTypeId As String, ByVal ParameterConfig As String, ByVal ParameterOrder As Integer) As Integer
            Return CType(SqlHelper.ExecuteScalar(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "UpdateParameter", ReportSetId, ParameterId, ParameterName, ParameterCaption, _
                ParameterTypeId, ParameterConfig, ParameterOrder), Integer)
        End Function
        Public Overrides Function ListParameter(ByVal ReportSetId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "ListParameter", ReportSetId), IDataReader)
        End Function

        Public Overrides Sub UpdateParameterOrder(ByVal ParameterId As Integer, ByVal Increment As Integer)
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "UpdateParameterOrder", ParameterId, Increment)
        End Sub
        Public Overrides Sub DeleteParameter(ByVal ParameterId As Integer)
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "DeleteParameter", ParameterId)
        End Sub

        ' report type
        Public Overrides Function ListReportType() As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "ListReportType"), IDataReader)
        End Function

        Public Overrides Function GetReportType(ByVal ReportTypeId As String) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetReportType", ReportTypeId), IDataReader)
        End Function

        ' Parameter type
        Public Overrides Function ListParameterType() As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "ListParameterType"), IDataReader)
        End Function

        Public Overrides Function GetParameterType(ByVal ParameterTypeId As String) As IDataReader
            Return CType(SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetParameterType", ParameterTypeId), IDataReader)
        End Function

        ' Connection
        Public Overrides Function GetConnection(ByVal ConnectionId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(Me.ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "GetConnection", ConnectionId), IDataReader)
        End Function

        Public Overrides Function UpdateConnection(ByVal PortalId As Integer, ByVal ConnectionId As Integer, ByVal ConnectionName As String, ByVal ConnectionString As String) As Integer
            Return CType(SqlHelper.ExecuteScalar(Me.ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "UpdateConnection", PortalId, ConnectionId, ConnectionName, ConnectionString), Integer)
        End Function
        Public Overrides Sub DeleteConnection(ByVal ConnectionId As Integer)
            SqlHelper.ExecuteNonQuery(Me.ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "DeleteConnection", ConnectionId)
        End Sub
        Public Overrides Function ListConnection(ByVal PortalId As Integer) As IDataReader
            Return CType(SqlHelper.ExecuteReader(Me.ConnectionString, DatabaseOwner & ObjectQualifier & CmdPrefix & "ListConnection", PortalId), IDataReader)
        End Function
    End Class

End Namespace

File: Providers/SqlDataProvider/SqlDataProvider.cs
using System;
using System.Data;
using System.Data.SqlClient;
using System.Text.RegularExpressions;
using DotNetNuke.Common.Utilities;
using DotNetNuke.Framework.Providers;
using Microsoft.ApplicationBlocks.Data;

namespace DNNStuff.SQLViewPro
{
    public class SqlDataProvider : DataProvider
    {
        #region Constructors

        public SqlDataProvider()
        {
            // Read the configuration specific information for this provider
            var providerConfiguration = ProviderConfiguration.GetProviderConfiguration("data");
            var objProvider = (Provider) (providerConfiguration.Providers[providerConfiguration.DefaultProvider]);

            // Read the attributes for this provider
            ConnectionString = Config.GetConnectionString();

            ProviderPath = objProvider.Attributes["providerPath"];

            ObjectQualifier = objProvider.Attributes["objectQualifier"];
            if (ObjectQualifier != "" && ObjectQualifier.EndsWith("_") == false)
            {
                ObjectQualifier += "_";
            }

            DatabaseOwner = objProvider.Attributes["databaseOwner"];
            if (DatabaseOwner != "" && DatabaseOwner.EndsWith(".") == false)
            {
                DatabaseOwner += ".";
            }
        }

        public string ConnectionString { get; }

        public string ProviderPath { get; }

        public string ObjectQualifier { get; }

        public string DatabaseOwner { get; }

        public string CmdPrefix => "DNNStuff_SQLViewPro_";

        #endregion

        public override DataSet RunQuery(string queryText, string dataSetName)
        {
            // make replacements for objectQualifier and databaseOwner
            queryText = Regex.Replace(queryText, "{oQ}|{objectQualifier}", ObjectQualifier, RegexOptions.IgnoreCase);
            queryText = Regex.Replace(queryText, "{dO}|{databaseOwner}", DatabaseOwner, RegexOptions.IgnoreCase);
            using (var cn = new SqlConnection(ConnectionString))
            {
                using (var cmd = SqlHelper.CreateCommand(cn, queryText, null))
                {
                    cmd.CommandTimeout = 0;
                    cmd.CommandType = CommandType.Text;
                    var adapter = new SqlDataAdapter(cmd);
                    cn.Open();
                    var ds = new DataSet(dataSetName);
                    adapter.Fill(ds);
                    return ds;
                }
            }
        }

        // report set
        public override IDataReader GetReportSet(int reportSetId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetReportSet", reportSetId);
        }

        public override IDataReader GetReportSetByModule(int moduleId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetReportSetByModule", moduleId);
        }

        public override int UpdateReportSet(int moduleId, int reportSetId, string reportSetName, string reportSetTheme,
            int reportSetConnectionId, string reportSetHeaderText, string reportSetFooterText, bool reportSetDebug,
            string runCaption, string backCaption, string parameterLayout, bool alwaysShowParameters, string renderMode,
            bool autoRun, string reportSetConfig)
        {
            return
                Convert.ToInt32(SqlHelper.ExecuteScalar(ConnectionString,
                    DatabaseOwner + ObjectQualifier + CmdPrefix + "UpdateReportSet", moduleId, reportSetId,
                    reportSetName, reportSetTheme, reportSetConnectionId, reportSetHeaderText, reportSetFooterText,
                    reportSetDebug, runCaption, backCaption, parameterLayout, alwaysShowParameters, renderMode, autoRun,
                    reportSetConfig));
        }

        public override IDataReader ListReportSet(int portalId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "ListReportSet", portalId);
        }

        public override void DeleteReportSet(int moduleId)
        {
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner + ObjectQualifier + CmdPrefix + "DeleteReportSet",
                moduleId);
        }

        public override IDataReader GetReportSetReport(int reportSetId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetReportSetReport", reportSetId);
        }

        public override IDataReader GetReportSetParameter(int reportSetId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetReportSetParameter", reportSetId);
        }

        // report
        public override IDataReader GetReport(int reportId)
        {
            return SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner + ObjectQualifier + CmdPrefix + "GetReport",
                reportId);
        }

        public override int UpdateReport(int reportSetId, int reportId, string reportTypeId, string reportName,
            string reportTheme, int reportConnectionId, string reportHeaderText, string reportFooterText,
            string reportCommand, string reportConfig, int reportOrder, int reportDrilldownReportId,
            string reportDrilldownFieldname, string reportNoItemsText, string reportPageTitle,
            int reportCommandCacheTimeout, string reportMetaDescription, string reportCommandCacheScheme)
        {
            return
                Convert.ToInt32(SqlHelper.ExecuteScalar(ConnectionString,
                    DatabaseOwner + ObjectQualifier + CmdPrefix + "UpdateReport", reportSetId, reportId, reportTypeId,
                    reportName, reportTheme, reportConnectionId, reportHeaderText, reportFooterText, reportCommand,
                    reportConfig, reportOrder, reportDrilldownReportId, reportDrilldownFieldname, reportNoItemsText,
                    reportPageTitle, reportCommandCacheTimeout, reportMetaDescription, reportCommandCacheScheme));
        }

        public override IDataReader ListReport(int reportSetId)
        {
            return SqlHelper.ExecuteReader(ConnectionString, DatabaseOwner + ObjectQualifier + CmdPrefix + "ListReport",
                reportSetId);
        }

        public override IDataReader ListReportDrilldown(int reportId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "ListReportDrilldown", reportId);
        }

        public override void UpdateReportOrder(int reportId, int increment)
        {
            SqlHelper.ExecuteNonQuery(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "UpdateReportOrder", reportId, increment);
        }

        public override void DeleteReport(int reportId)
        {
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner + ObjectQualifier + CmdPrefix + "DeleteReport",
                reportId);
        }

        // Parameter
        public override IDataReader GetParameter(int parameterId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetParameter", parameterId);
        }

        public override int UpdateParameter(int reportSetId, int parameterId, string parameterName,
            string parameterCaption, string parameterTypeId, string parameterConfig, int parameterOrder)
        {
            return
                Convert.ToInt32(SqlHelper.ExecuteScalar(ConnectionString,
                    DatabaseOwner + ObjectQualifier + CmdPrefix + "UpdateParameter", reportSetId, parameterId,
                    parameterName, parameterCaption, parameterTypeId, parameterConfig, parameterOrder));
        }

        public override IDataReader ListParameter(int reportSetId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "ListParameter", reportSetId);
        }

        public override void UpdateParameterOrder(int parameterId, int increment)
        {
            SqlHelper.ExecuteNonQuery(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "UpdateParameterOrder", parameterId, increment);
        }

        public override void DeleteParameter(int parameterId)
        {
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner + ObjectQualifier + CmdPrefix + "DeleteParameter",
                parameterId);
        }

        // report type
        public override IDataReader ListReportType()
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "ListReportType");
        }

        public override IDataReader GetReportType(string reportTypeId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetReportType", reportTypeId);
        }

        // Parameter type
        public override IDataReader ListParameterType()
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "ListParameterType");
        }

        public override IDataReader GetParameterType(string parameterTypeId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetParameterType", parameterTypeId);
        }

        // Connection
        public override IDataReader GetConnection(int connectionId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "GetConnection", connectionId);
        }

        public override int UpdateConnection(int portalId, int connectionId, string connectionName,
            string connectionString)
        {
            return
                Convert.ToInt32(SqlHelper.ExecuteScalar(ConnectionString,
                    DatabaseOwner + ObjectQualifier + CmdPrefix + "UpdateConnection", portalId, connectionId,
                    connectionName, connectionString));
        }

        public override void DeleteConnection(int connectionId)
        {
            SqlHelper.ExecuteNonQuery(ConnectionString, DatabaseOwner + ObjectQualifier + CmdPrefix + "DeleteConnection",
                connectionId);
        }

        public override IDataReader ListConnection(int portalId)
        {
            return SqlHelper.ExecuteReader(ConnectionString,
                DatabaseOwner + ObjectQualifier + CmdPrefix + "ListConnection", portalId);
        }
    }
}


